File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/D-Таблица_умножения/solution.py
import sys


def get_le(middle: int, n: int) -> int:
    le: int = 0
    for num in range(1, n + 1):
        le += min(middle // num, n)
    return le


def solution() -> None:
    """
    Time Complexity O(Nlog(N^2))
    Memory Complexity O(1)
    """
    n, k = map(int, sys.stdin.readline().rstrip().split())
    left: int = 1
    right: int = n * n
    while left <= right:
        middle: int = left + ((right - left) >> 1)
        if get_le(middle, n) < k: left = middle + 1 
        else: right = middle - 1 
    sys.stdout.write('{}\n'.format(left))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/D-Таблица_умножения/solution.cpp
#include <iostream>


long long getLE(long long& middle, long long& n) {
    long long le = 0;
    for (long long num = 1; num <= n; ++num) {
        le += std::min(middle / num, n);
    }
    return le;
}


void solution() {
    /*
    Time Complexity O(Nlog(N^2))
    Memory Complexity O(1)
    */
    long long n, k;
    std::scanf("%lld %lld", &n, &k);
    long long left = 1;
    long long right = n * n;
    while (left <= right) {
        long long middle = left + (right - left) / 2;
        long long lessThanOrEqual = getLE(middle, n);
        if (lessThanOrEqual < k) {
            left = middle + 1;
        } else right = middle - 1;
    }
    std::printf("%lld\n", left);
}


int main() {
    solution();
}